### ISSUE TITLE
[$750] Selecting and  copy / pasting multiple messages doesn't work as expected

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#37447]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.44-11
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @Beamanator 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1709113646585099

## Action Performed:

1. Send 3 messages
2. Try to select all 3 messages at the same time -> This is not easy unless you start by selecting the right side of a message & dragging down, which isn't expected
3. Copy and paste in the compose box

If you copy & paste at that point, the pasted messages show up in the wrong order

## Expected Result:
Should retain the order of the messages

## Actual Result:
unknown

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

https://github.com/Expensify/App/assets/38435837/c0db8ba3-36d8-417f-b886-d55da9397c7b

https://github.com/Expensify/App/assets/38435837/de8f01f8-d282-4078-9c8c-545060e039a0

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~015529b5b6c53de2d3</li>
        <li>Upwork Job ID: 1762895944035651584</li>
        <li>Last Price Increase: 2026-02-26</li>
<li>Automatic offers: </li>
<ul>
<li>NJ-2020 | Contributor | 110427963</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @neerajbachani</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @neerajbachani</details>

**Comment by Christinadobrzyn** [comment]
I think this is going to be fixed with https://github.com/Expensify/App/issues/35946

Going to put this on hold for that GH and we can reevaluate if it's still happening after.

**Comment by Christinadobrzyn** [comment]
monitoring https://github.com/Expensify/App/issues/35946

**Comment by Christinadobrzyn** [comment]
monitoring https://github.com/Expensify/App/issues/35946

**Comment by Christinadobrzyn** [comment]
monitoring https://github.com/Expensify/App/issues/35946

**Comment by Christinadobrzyn** [comment]
monitoring https://github.com/Expensify/App/issues/35946

**Comment by Christinadobrzyn** [comment]
monitoring https://github.com/Expensify/App/issues/35946

**Comment by Christinadobrzyn** [comment]
https://github.com/Expensify/App/issues/35946

**Comment by Christinadobrzyn** [comment]
hold for https://github.com/Expensify/App/issues/35946

**Comment by Christinadobrzyn** [comment]
hold for https://github.com/Expensify/App/issues/35946

**Comment by Christinadobrzyn** [comment]
I think we're still on hold for - https://github.com/Expensify/App/issues/33725

**Comment by Christinadobrzyn** [comment]
I think we're still on hold for - https://github.com/Expensify/App/issues/33725 

@eh2077 do you think it makes sense to keep this on hold for https://github.com/Expensify/App/issues/33725

**Comment by eh2077** [comment]
@Christinadobrzyn Yes, I think they're very likely to share same root cause, so makes sense to put it on hold

**Comment by Christinadobrzyn** [comment]
still on hold for - https://github.com/Expensify/App/issues/33725

**Comment by Christinadobrzyn** [comment]
@eh2077 should we still keep this on hold for https://github.com/Expensify/App/issues/33725?

**Comment by eh2077** [comment]
@Christinadobrzyn yes I think so

**Comment by Christinadobrzyn** [comment]
on hold for https://github.com/Expensify/App/issues/33725

Just a heads up - I'm going to be ooo until June 24th. I'm not going to assign anyone new but if you need a new BZ teammate for any reason please feel free to ask for one to be assigned here.

**Comment by Christinadobrzyn** [comment]
on hold for https://github.com/Expensify/App/issues/33725

**Comment by Christinadobrzyn** [comment]
https://github.com/Expensify/App/issues/33725

**Comment by Christinadobrzyn** [comment]
on hold for https://github.com/Expensify/App/issues/33725

**Comment by Christinadobrzyn** [comment]
on hold for https://github.com/Expensify/App/issues/33725

**Comment by Christinadobrzyn** [comment]
on hold for https://github.com/Expensify/App/issues/33725

**Comment by Christinadobrzyn** [comment]
on hold for https://github.com/Expensify/App/issues/33725

**Comment by Christinadobrzyn** [comment]
on hold for https://github.com/Expensify/App/issues/33725

**Comment by Christinadobrzyn** [comment]
on hold for https://github.com/Expensify/App/issues/33725

**Comment by Christinadobrzyn** [comment]
on hold for https://github.com/Expensify/App/issues/33725

**Comment by Christinadobrzyn** [comment]
on hold for https://github.com/Expensify/App/issues/33725

**Comment by Christinadobrzyn** [comment]
on hold for https://github.com/Expensify/App/issues/33725

**Comment by Christinadobrzyn** [comment]
on hold for https://github.com/Expensify/App/issues/33725

**Comment by Christinadobrzyn** [comment]
https://github.com/Expensify/App/issues/33725 is closed @eh2077 should we test again?

Just a heads up that I'm going to be ooo May 21st - May 27th. Back on the 28th. 

I'm not going to assign another BZ teammate but if you need someone while I'm gone, please remove and add the Bug label.

**Comment by eh2077** [comment]
@Christinadobrzyn Yeah, it'll be great to ask QA to retest the bug.

**Comment by Christinadobrzyn** [comment]
asking to retest here - https://expensify.slack.com/archives/C9YU7BX5M/p1748553627871799

**Comment by eh2077** [comment]
@mvtglobally I think there're at least two issues here.

### Issue 1: Selecting multiple messages doesn't work. You can't select multiple messages from top-left to bottom-right.

https://github.com/user-attachments/assets/179691bc-aa8a-4bd0-b216-ba6bbbf3baa8

Instead, you have to select multiple messages from top-right to left-bottom, which should be unexpected.

https://github.com/user-attachments/assets/7af2074a-459d-461b-9010-3cd2805d8522

---

### Issue 2: Copy and paste multiple messages shows in reverse order

https://github.com/user-attachments/assets/af590b2c-7dd7-467d-ae50-c556d45931d7

---

I think the two issues are very likely to share same root cause, so I think we should tackle them together. Clarifying them separately makes it easier to understand the issue and the expected behaviors.

@mvtglobally @Christinadobrzyn Do you agree?

**Comment by Christinadobrzyn** [comment]
Hi @eh2077 I think this is a good plan! Do you think we can use this GH to resolve the two issues?

**Comment by eh2077** [comment]
@Christinadobrzyn Yes, I think so. Actually, the original bug report mentions the two issues.

**Comment by Christinadobrzyn** [comment]
awesome! sounds good! So I think we're waiting on proposals.

**Comment by nabi-ebrahimi** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Selecting and copy / pasting multiple messages doesn't work as expected

### What is the root cause of that problem?

We already had this issue in 2021 #3381 and solved it by these Prs

[pr1](https://github.com/Expensify/react-native-web/pull/3), [pr2](https://github.com/Expensify/App/pull/6738), [pr3](https://github.com/Expensify/App/pull/6798)

### What changes do you think we should make in order to solve the problem?

We need to reapply those changes to make it work. The components and code have changed significantly, but I believe the same adjustments will resolve the issue.

### What specific scenarios should we cover in automated tests to prevent reintroducing this issue in the future?

None

### What alternative solutions did you explore? (Optional)

**Comment by nabi-ebrahimi** [comment]
I am trying to find the offending PR

**Comment by nabi-ebrahimi** [comment]
Not sure, this can be the offending PR #24482

**Comment by eh2077** [comment]
@m-natarajan Thanks for your proposal! Can you elaborate on the root cause in your proposal? The PRs you linked are very old and it's not easy to understand the root cause by looking them.

**Comment by nabi-ebrahimi** [comment]
This is the offending PR. #24482 I reverted the changes related to `InvertedFlatList`, and it's working again.

https://github.com/user-attachments/assets/0c66b8ba-0679-4195-b043-ea11d89e0469

**Comment by nabi-ebrahimi** [comment]
> [@m-natarajan](https://github.com/m-natarajan) Thanks for your proposal! Can you elaborate on the root cause in your proposal? The PRs you linked are very old and it's not easy to understand the root cause by looking them.

@eh2077 I think you tagged @m-natarajan by mistake.

This issue was already addressed in the previous PRs. However, when they upgraded `react-native-web` [here](#24482), they failed to properly apply those changes.

**Comment by nabi-ebrahimi** [comment]
@eh2077 , @Christinadobrzyn, I’ll review the existing patch for this, and if necessary, create a new one to include the missing changes. That should resolve the issue.

**Comment by eh2077** [comment]
Still looking for a decent proposal!

**Comment by Christinadobrzyn** [comment]
I'll see if I can get someone from SWM to look at this. https://expensify.slack.com/archives/C08JJ67N602/p1749592670921119

**Comment by blazejkustra** [comment]
Hey @Christinadobrzyn, I can take a look later this week! I’ll see what we can do about it (though I assume it won’t be easy 😅)

**Comment by Christinadobrzyn** [comment]
HI @blazejkustra, amazing! Thank you! I assigned you to this GH for when you're available.

**Comment by blazejkustra** [comment]
Hey, just a quick update from my side. I can confirm that what @nabi-ebrahimi suggested is correct.

TL;DR: This issue originally came up in 2021 and was eventually fixed (the [bounty for it was huge](https://github.com/Expensify/App/issues/3381#issuecomment-1002286715) a total of $20.5k!). The fix was applied on a fork of `react-native-web` and then used in NewDot.

~In late 2023, @getusha upgraded `react-native-web` to the latest version in [this PR](https://github.com/Expensify/App/pull/24482)~, which unfortunately dropped some of the custom patches from the fork, including the one for this issue. That’s probably why it resurfaced, though it slipped through unnoticed at the time 🫠

edit: It wasn't @getusha's PR, it was [this](https://github.com/Expensify/App/pull/32098#discussion_r2151719509).

After applying the old fix the copy / pasting functionality works again: 

https://github.com/user-attachments/assets/c5e0de99-7ed2-4d81-b40f-bbbdd9f03997

@Christinadobrzyn @nabi-ebrahimi I think it would be fair for @nabi-ebrahimi to continue work on this issue as he was first to propose the correct solution. What do you think?

**Comment by getusha** [comment]
Let me know if there is anything I can do to help 😃

**Comment by blazejkustra** [comment]
> Let me know if there is anything I can do to help 😃

Just one question @getusha, any idea why those patches didn’t get copied over? Maybe it was intentional?

**Comment by getusha** [comment]
@blazejkustra It was initially a fork, we abandoned the fork and made a patch https://github.com/Expensify/App/pull/24482/files#diff-a70cbe6e9f9b62da64c873dda7d94a84227d843bb165a6f283148b1cfdb2d09b

It's possible we missed something and the patch isn't working.

~~It's been a year though, It's possible the upgrade didn't cause it.~~

**Comment by blazejkustra** [comment]
> It's been a year though, It's possible the upgrade didn't cause it.

I think you are right @getusha, it wasn't your PR. I found the exact [PR](https://github.com/Expensify/App/pull/32098/files#r2151719509) that removed this patch and caused copy / pasting issue. 

@perunt could you share some insights why the code was removed?

**Comment by perunt** [comment]
Hey, this change was made by @janicduplessis, but as far as I remember, we rewrote the scroll component and removed that hack with inversion since it was no longer needed

**Comment by blazejkustra** [comment]
Thanks @perunt! @janicduplessis, any chance you remember more about why it was removed? I understand that the scroll component was rewritten, but this 'hack' was a [$20k fix](https://github.com/Expensify/App/issues/37447#issuecomment-2977022833) and it might still be needed 😅

**Comment by blazejkustra** [comment]
I'm waiting on updates from @janicduplessis and perhaps [migration to legend list](https://expensify.slack.com/archives/C05LX9D6E07/p1751274223903399?thread_ts=1748529627.357659&cid=C05LX9D6E07) might fix this as well

**Comment by blazejkustra** [comment]
Turns out migration to LegendList potentially fixes this issue, so I'm just going to wait for now 👍 

https://github.com/user-attachments/assets/5851fe9a-dbd4-42c1-8cde-18aa42e8415f

**Comment by Christinadobrzyn** [comment]
Oh nice! @blazejkustra do you think this is resolved then or are we waiting on something?

**Comment by Christinadobrzyn** [comment]
Update for Melvin - we're working on the proposal

**Comment by blazejkustra** [comment]
@Christinadobrzyn Shall we make it weekly or even monthly?

**Comment by Christinadobrzyn** [comment]
working on a proposal right @blazejkustra @eh2077?

**Comment by blazejkustra** [comment]
I was hoping that @zirgulis’s proposal to migrate chat to the legend list would resolve this issue as well, but it seems there’s been no progress [there](https://expensify.slack.com/archives/C05LX9D6E07/p1751975257799699?thread_ts=1748529627.357659&cid=C05LX9D6E07). 

I’ll try migrating chat to a different library (FlashList), which we’re already using in the codebase, and will share an update next week. I’m waiting for it to be upgraded [to the newest version](https://github.com/Expensify/App/pull/67745), after which I’ll run some tests.

**Comment by blazejkustra** [comment]
We’re still waiting for a fix in the upstream repository, so no changes on our side yet.

**Comment by nabi-ebrahimi** [comment]
Hi, Is this open for external contributors? I’d be happy to work on fixing it in the upstream repo.

**Comment by blazejkustra** [comment]
I don’t see any issues with contributors submitting new proposals. I’ll be glad to help with the review. What do you think? @Christinadobrzyn @eh2077

Just be aware @NJ-2020 @nabi-ebrahimi that this won't be easy to fix 😄

**Comment by NJ-2020** [comment]
Sorry for the delay

<hr/>

> `@NJ-2020` meanwhile, could you also try using legend-list v2.1.0-beta.0? I reached out to the author and he replied [here](https://x.com/jmeistrich/status/1975569915768971380)

@blazejkustra Sure, thanks, I'll look into this soon

**Comment by NJ-2020** [comment]
Hi 👋 Sharing an update

<hr />

Why legend-list v2.1.0-beta.0 works properly when selecting multiple line of texts after scrolling?, it's because:

- When displaying the lists, we have custom hook called `useDOMOrder` which will re-order the list in the DOM tree based on the index item not key

https://github.com/LegendApp/legend-list/blob/073d67ad81ca3287b0eb973cadd1053573bba2d4/src/components/Containers.tsx#L32-L33

So from:
```
[5,2,3,4]
keyMap = Map{
  "1": { index: 5, itemType: "default", stableId: "item-5" },
  "2": { index: 2, itemType: "default", stableId: "item-2" },
  "3": { index: 3, itemType: "default", stableId: "item-3" },
  "4": { index: 4, itemType: "default", stableId: "item-4" },
}
```
Will become:
```
[2,3,4,5]
keyMap = Map{
  "2": { index: 2, itemType: "default", stableId: "item-2" },
  "3": { index: 3, itemType: "default", stableId: "item-3" },
  "4": { index: 4, itemType: "default", stableId: "item-4" },
  "1": { index: 5, itemType: "default", stableId: "item-5" },
}
```
- Update/sort/move DOM element based on the `index` attribute value 
https://github.com/LegendApp/legend-list/blob/web-beta0/src/hooks/useDOMOrder.ts
https://github.com/LegendApp/legend-list/blob/web-beta0/src/utils/reordering.ts
- Pass `index` value to each container
https://github.com/LegendApp/legend-list/blob/073d67ad81ca3287b0eb973cadd1053573bba2d4/src/components/Container.tsx#L175

**Comment by NJ-2020** [comment]
@eh2077 @blazejkustra Please kindly review above ^^ when you got a chance 🧑‍💻

Many thanks

**Comment by Christinadobrzyn** [comment]
moving to weekly for review - @eh2077 and @blazejkustra can you please review when possible? TY!

**Comment by eh2077** [comment]
Reviewing

**Comment by NJ-2020** [comment]
Hi 👋 Apologies for the delay, I've been busy with some issues over the past few days

Update: I'm currently investigating the current `react-native-web` **VirtualizedList** implementation and will share a complete proposal soon.

cc: @eh2077 

Many thanks.

**Comment by Christinadobrzyn** [comment]
Thanks for the update @NJ-2020!

**Comment by NJ-2020** [comment]
**Update:**

I am still working on it, I should have the complete proposal soon. Thanks

**Comment by eh2077** [comment]
Thanks! Looking into it.

**Comment by NJ-2020** [comment]
Hi 👋 Apologies for the delay, I'll share an update (regarding [this](https://github.com/Expensify/App/issues/37447#issuecomment-3526922537)) / or complete proposal soon, in the meanwhile, here's quick short demo comparison if we use `pushOrUnshift` along with `MVCP` enabled:

**Left** side is **Chrome**, and the right side is **Safari**
**Main branch**, normal push  | **[fix/37447-1](https://github.com/NJ-2020/Expensify/tree/fix/37447-1) branch** using `pushOrUnshift`
------------- | -------------
<video src="https://github.com/user-attachments/assets/a26e6c54-a923-41dd-843e-e40f37d150c0"></video>  | <video src="https://github.com/user-attachments/assets/28df4661-3d76-47ba-9271-36dba6c0b777"></video>

As we can see here, on **Safari** when using `MVCP` enabled and using `pushOrUnshift` doesn't work well, because `MVCP` works by retrieving the cell & data item, so if the data lists is `[1,2,3]` then inside the cells it got reverted by using `pushOrUnshift` which has misconception index/item value which causes `MVCP` to not work properly in this case.

I will share more details/update soon.

<hr />

FYI: `MVCP` implementation was [added](https://github.com/necolas/react-native-web/pull/2588#issue-1927148633) to specifically also to fix the Safari scroll lists bug
> Currently the behavior varies depending on the browser. In Chrome the content position is maintained if the scroll position is > 0, while on Safari it is never maintained. The Chrome behavior diverges from react-native so this also makes sure it is consistent, which is that by default the content position is not maintained, and that when the maintainVisibleContentPosition prop is set it is.

Many thanks.

cc: @blazejkustra @eh2077

**Comment by NJ-2020** [comment]
Hi 👋, sharing an **updates** below 👇

**Comment by NJ-2020** [comment]
## **Why `pushOrUnshift` are being removed in this [PR](https://github.com/Expensify/App/pull/32098)?**

### Context:
Before going deep dive into the root cause, let's see why/how `maintainVisibleContentPosition` is being implemented/works
<hr />

So when new items are added (deeplink comment, new message loaded/arrived) browsers behave differently:

1. **Safari**: Scroll is never maintained, position jumps - content shifts down
2. **Chrome**: The content position is maintained if the scroll position is > 0 / (when not scrolled to the top, i.e it middle)

Reference:
> Currently the behavior varies depending on the browser. In Chrome the content position is maintained if the scroll position is > 0, while on Safari it is never maintained. The Chrome behavior diverges from react-native so this also makes sure it is consistent, which is that by default the content position is not maintained, and that when the maintainVisibleContentPosition prop is set it is.

By having `MVCP` implemented/enabled, it will allow us to keep/maintain the correct scroll position when new items are added.

`MVCP` works by **2 layers**:
1. First layer ([react-native-web patch](https://github.com/Expensify/App/blob/main/patches/react-native-web/react-native-web%2B0.21.2%2B001%2Binitial.patch)): Make sure the correct items (DOM) are rendered when data changes
2. Second layer ([src/components/FlatList/index.tsx](https://github.com/Expensify/App/blob/938fac59abc30bc1bfc1b2a2ad8c2eda402397f3/src/components/FlatList/index.tsx#L92-L190)): Maintain the correct scroll position, when new items are added, `FlatList`/`VirtualizedList` uses spacer as the remaining message i.e (`data.length` = `100`, `renderedItems` = `10`, `spacer` = `90`)

### First layer:

**Simple example:**
```
data = [3,4,5,6,7]
maximumRenderedItems = 5

firstVisibleItem = 7; index = [4]

// Track which item need to be rendered (first, last)
cellsAroundViewport = {
    "first": 0,
    "last": 4
}

// Rendered items based on cellsAroundViewport
domItems = [3,4,5,6,7]
```

**New message arrives (loaded) `1,2`**:
```
data = [1,2,3,4,5,6,7]
```

**Without MVCP enabled**, it would display to the user / render (DOM):
```
domItems = [1,2,3,4,5]
```
Because `cellsAroundViewport` tracks the data by index, so when the lists changes, it would still use the same index
```
cellsAroundViewport = {
    "first": 0,
    "last": 4
}
```

 **With MVCP enabled**, it will automatically adjust the cellsAroundViewport, by updating with new index (find the new index based on firstVisibleItem on the data lists which is **6**)
```
// Find firstVisibleItem in data, found at index 6,
domItems = [3,4,5,6,7]
```

But inside VirtualizedList/FlatList, we have something called **spacer**, spacer is a View component used for un-rendered items, i.e
```tsx
data = [1,2,3,4,5,6,7,8] // Note: this is non-inverted data lists, in the visual order, it will become: 8,7,6...,etc
maximumRenderedItems = 3;

// Assume each item height is 20px
itemHeight = "20px"

// Rendered items
domItems = [1,2,3]

// Create (top) spacer for the remaining items 4,5,6,7,8 = 20px * 5 = 100px
spacer = <View ... style={{height: "50px}} ...>...</View>

```

So even if we already adjust the the rendered DOM items (**cellsAroundViewport**), the scroll offset is still changing due to the spacer

So we fix this by adding another layer, which is Layer 2

### Second layer:
The second layer tracks the first visible item top offset (distance between the item & parent border corner, so offsetTop can changed depending on the message length/height) above/after > the scroll offset, so the first visible item to the user
https://github.com/Expensify/App/blob/900ba48c8cab5f5448a9d3c104b3d75916b80c77/src/components/FlatList/index.tsx#L92-L115

<img width="1920" height="954" alt="Image" src="https://github.com/user-attachments/assets/c018ef8f-1967-42a4-9647-2a0248090c48" />

So when new messages is arrived, the spacer will adjust it height and the layer 2 (scroll logic) will automatically adjust the scroll offset too
https://github.com/Expensify/App/blob/900ba48c8cab5f5448a9d3c104b3d75916b80c77/src/components/FlatList/index.tsx#L180-L181

<hr/>

### RCA/Why they remove it?
After looking into the PR, it seems they [removed](https://github.com/Expensify/App/pull/32098#issuecomment-1866705890) it because it causes this scroll bug [issue](https://github.com/Expensify/App/pull/32098#issuecomment-1845564943), https://github.com/Expensify/App/pull/32098#issuecomment-1845825197

I've tried apply this [changes](https://github.com/NJ-2020/Expensify/tree/fix/37447-1) and tried to reproduce (on **Safari** specifically) this [bug](https://github.com/Expensify/App/pull/32098#issuecomment-1845564943), but I couldn't reproduce it, i've tried many times but still I couldn't:


| Main chat testing (many test cases)  | Deeplink (comment) after clearing cache |
| ------------- | ------------- |
| <video src="https://github.com/user-attachments/assets/f5205617-e9ff-4aa0-a2e6-17b72488d1c2" />  | <video src="https://github.com/user-attachments/assets/e0534656-3fbe-4887-b730-91867ee12a23" />  |

**Note for (Deeplinking comment):** The flash scroll is also happened in staging/main

Maybe for deeper root cause, we can ask `@janicduplessis` why the `pushOrUnshift` logic specifically is being [removed](https://github.com/Expensify/App/pull/32098#issuecomment-1866705890) as the scroll bug [issue](https://github.com/Expensify/App/pull/32098#issuecomment-1845564943) does not reproducible anymore.

**Comment by NJ-2020** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Selecting and copy / pasting multiple messages doesn't work as expected
### What is the root cause of that problem?
- [Shopfiy/FlashList #1839](https://github.com/Shopify/flash-list/issues/1839): [RCA](https://github.com/Expensify/App/issues/37447#issuecomment-3396686049)
- [LegendApp/LegendList](https://github.com/LegendApp/legend-list/): [RCA](https://github.com/Expensify/App/issues/37447#issuecomment-3396686049)
- Why `legend-list v2.1.0-beta.0` works properly?: [RCA](https://github.com/Expensify/App/issues/37447#issuecomment-3417762583)
- Current [codebase FlatList](https://github.com/Expensify/App/blob/main/src/components/FlatList/index.tsx) / [react-native-web](https://github.com/necolas/react-native-web/): [RCA/Solution](https://github.com/Expensify/App/issues/37447#issuecomment-3526784927)
- Why `pushOrUnshift` is being removed in this [PR](https://github.com/Expensify/App/pull/32098)?: [RCA]()
### What changes do you think we should make in order to solve the problem?
Re-implement this [PR](https://github.com/Expensify/react-native-web/pull/2) and this https://github.com/Expensify/react-native-web/pull/3

But we need to modify the MVCP a bit, instead of tracking it in forwards direction, we need to change it to backward direction
https://github.com/Expensify/App/blob/836f696905423b18607733d3a755b01b010fb150/src/components/FlatList/index.tsx#L102-L111
```tsx
const prepareForMaintainVisibleContentPosition = useCallback(() => {
   //  ...
    
    const contentViewLength = contentView.childNodes.length;
    for (let i = mvcpMinIndexForVisible; i < contentViewLength; i++) {
        const subview = contentView.childNodes[restProps.inverted ? contentViewLength - i - 1 : i] as HTMLElement;
        const subviewOffset = horizontal ? subview.offsetLeft : subview.offsetTop;
        if (subviewOffset > scrollOffset) {
            prevFirstVisibleOffsetRef.current = subviewOffset;
            firstVisibleViewRef.current = subview;
            break;
        }
    }
}, [getContentView, getScrollOffset, mvcpMinIndexForVisible, horizontal, restProps.inverted]);
```

Because the current implementation (**MVCP**) is trying to track the first (latest visible) item at position `#2`
```
data = [1,2,3,4,5]
renderedItems = [1,2,3,4,5]
visualDisplay = [5,4,3,2,1] // reversed by scaleY(-1)

// **firstVisibleItem** = renderedItems (individual item) > above scroll offset (0) so > `1` - `mvcpMinIndexForVisible` >> `2`
firstVisibleItem = 2
```
So when `pushOrUnshift` is being used, it will reverse the rendered DOM items causing **MVCP** to track the oldest item (which is incorrect)

### [TEST BRANCH](https://github.com/NJ-2020/Expensify/tree/fix/37447-1)

### Result

https://github.com/user-attachments/assets/43c6be22-c5ea-4ef3-9565-47906d41c060

### What alternative solutions did you explore? (Optional)

**Comment by eh2077** [comment]
Looking into it

**Comment by NJ-2020** [comment]
@eh2077 Can you please provide an update?. Thanks

**Comment by Julesssss** [comment]
Okay lets give it try, we'll just need to include the prior bug as test cases to be sure it's no longer occuring.

**Comment by NJ-2020** [comment]
> `@NJ-2020` I think this https://github.com/Expensify/App/issues/37447#issuecomment-3903829731 is the issue we need to fix in the PR because the copy and paste function doesn't work for the app. Hmm, I'm a little bit confused here, do you mean your solution doesn't work anymore? Btw, I don't think it makes sense to work / test your solution with older version of the codebase, instead, we should alway work with the latest main.

@eh2077 No, I don't say may solution doesn't work anymore. My solution fixes the selections bug, however copying any message in the report chat lists doesn't work, this happen both on main/staging & my-branch. So it appears there's some new issues from (**main**) regarding selection/copying message (Message Chat List). So I tried the old **main** version (before these changes), and it works well + my solution.

> I think the solution should work for desktop browsers (both Chrome and Safari).

Currently, in **main** branch, copying (selection list) both in Chrome & Safari doesn't work, selecting multiple messages only works in Chrome.

> This online tool might be helpful to inspect clipboard, https://evercoder.github.io/clipboard-inspector/

The problem is the text doesn't get copied, it's showing empty text value, same with this tool (I've tried it), and also now, if we want to select any message, we can't start from any angle, we need to start from the first character of the message to be able to do selection. 

cc: @Julesssss

**Comment by Julesssss** [comment]
If we can find the offending PR that should make our next step clearer.

**Comment by Julesssss** [comment]
@MelvinBot please familiarise yourself with this issue. Are you able to locate the PR which introduced the Safari regression?

**Comment by MelvinBot** [comment]
I've investigated the issue history and codebase to identify the PR(s) responsible for the Safari regression. There are actually **two layers** of regression here:

---

## Layer 1: The Original Selection/Copy-Paste Regression (Jan 2024)

**PR [#32098](https://github.com/Expensify/App/pull/32098)** ("HIGH: (Comment linking: step 2) WEB maintain visible content position"), merged **2024-01-09**, is the root cause of the original multi-message selection and copy/paste ordering problem.

This PR rewrote the `react-native-web` VirtualizedList patch to add `maintainVisibleContentPosition` support. In doing so, it **removed the `pushOrUnshift` DOM-reordering approach** (from the original [$20.5k fix](https://github.com/Expensify/App/issues/3381) in PRs [#6738](https://github.com/Expensify/App/pull/6738) and [#6798](https://github.com/Expensify/App/pull/6798)) and reverted back to using `transform: scaleY(-1)` for list inversion. This breaks multi-message selection because browsers select based on DOM order, not visual order. This is the regression that @NJ-2020's [proposal](https://github.com/Expensify/App/issues/37447#issuecomment-3630411970) addresses via a react-native-web patch.

---

## Layer 2: The Recent Regression (Dec 2025 - Feb 2026)

@NJ-2020 reports that **copying messages and Safari selection stopped working entirely** between ~Dec 9, 2025 and Feb 15, 2026 — even for single messages. This is a **separate, newer regression** on top of the original one.

After investigating commits touching the selection/clipboard pipeline, FlatList, react-native-web patches, and report action rendering during this window, I found that **no PR directly modified** `SelectionScraper`, `Clipboard`, `useCopySelectionHelper`, or the `userSelect` CSS styles. However, several PRs changed the re-rendering behavior of the chat list in ways that could destroy browser text selection:

### Primary Suspects

**1. PR [#77790](https://github.com/Expensify/App/pull/77790)** — Merged **Dec 30, 2025**
- **Removed `memo()` from `TextCommentFragment`**, meaning every parent re-render now causes the text DOM nodes to re-render. On web, re-rendering DOM subtrees that have active `window.getSelection()` ranges destroys those ranges. Safari is significantly more sensitive to this than Chrome.

**2. PR [#78634](https://github.com/Expensify/App/pull/78634) + [#81210](https://github.com/Expensify/App/pull/81210)** — "new Onyx Derived Value VISIBLE\_REPORT\_ACTIONS" — Merged Jan 22, reverted Jan 30, **re-applied Feb 2**
- Changed how the `reportActions` array feeding the FlatList is computed. If the derived value produces new array references more frequently, it triggers FlatList re-renders that destroy active browser selection state.

**3. PR [#76711](https://github.com/Expensify/App/pull/76711)** — Merged **Dec 22, 2025**
- Added `shouldHideContent` to the web FlatList, applying `visibility: hidden` to the content container during initial layout. If `shouldScrollToEndAfterLayout` gets stuck as `true`, content stays hidden and unselectable.

### Secondary Suspects

**4. PR [#80153](https://github.com/Expensify/App/pull/80153)** — Merged **Jan 27, 2026**
- Made `RenderHTML` selectability configurable; introduced `isSelectable={false}` for certain message types. Mixing selectable and non-selectable nodes in the same list can break Safari's selection engine.

---

## Recommended Next Steps

1. **Quick validation**: Re-add `memo()` to `TextCommentFragment` on the current `main` branch and test if selection/copying works again on Safari. If it does, the combination of unmemoized components + more frequent data updates is the root cause of the Layer 2 regression.

2. **Git bisect**: Bisect between Dec 9 (known working) and Dec 30 (first major suspect) to pinpoint the exact commit.

3. **Runtime logging**: Add `console.log` in `getHTMLOfSelection()` to check what `window.getSelection()` returns, and add render-count logs to `PureReportActionItem` and `TextCommentFragment` to verify if re-renders are happening during selection.

4. Once the Layer 2 regression is fixed, @NJ-2020's proposal to patch react-native-web should address the Layer 1 issue (multi-message selection ordering).

---
**Confidence**: High on Layer 1 (confirmed by @blazejkustra and code evidence). Medium on Layer 2 (multiple candidate PRs, no single "smoking gun" — a bisect would confirm).

**Next Steps for Contributor+ team:** Reply with `@MelvinBot implement this` to create a draft PR, or `@MelvinBot <your feedback>` to refine this analysis.

**Comment by NJ-2020** [comment]
I will provide an update today. Thanks

**Comment by Julesssss** [comment]
> Found the offending [PR](https://github.com/Expensify/App/pull/78527)

Good find. It's a bit late to revert, but we can increase the bounty by 50% if that can be fixed here.

**Comment by NJ-2020** [comment]
Hi 👋, I'll be OOO today 19 March for about 1~2 weeks. Thanks

cc: @Julesssss

**Comment by NJ-2020** [comment]
Hi 👋 Apologies for the delay

<hr />

[PR](https://github.com/Expensify/App/pull/85306) is ready.

Thanks

**Comment by mallenexpensify** [comment]
Please comment and tag me if/when payment is due, thx

**Comment by NJ-2020** [comment]
Hi 👋, This [PR](https://github.com/Expensify/App/pull/85114) has been merged (which migrate the current main chat from FlatList >> **FlashList**), and here's an open PR: https://github.com/Expensify/App/pull/85825 (to fix the FlashList rendering DOM items)

So I think we can only fix this [issue](https://github.com/Expensify/App/issues/37447#issuecomment-3903829731) for this [PR](https://github.com/Expensify/App/pull/85306), wdyt?. Thanks

cc: @eh2077 @Julesssss

**Comment by eh2077** [comment]
yes, I think so. Let's ensure the original issue get fixed and minimize hack in the patch.

**Comment by Julesssss** [comment]
Agree

**Comment by NJ-2020** [comment]
[PR](https://github.com/Expensify/App/pull/85306) is ready

<hr />


**Just a note**: After this [PR](https://github.com/Expensify/App/pull/85306) is merged, we'll need to wait for the other [PR](https://github.com/Expensify/App/pull/85825) to be merged to fully resolve this issue and re-test. Thanks


cc: @eh2077

**Comment by eh2077** [comment]
same as https://github.com/Expensify/App/issues/37447#issuecomment-4286638207

**Comment by NJ-2020** [comment]
Hi 👋, apologies for the delay

<hr />

I'll provide an update tomorrow.

**Comment by NJ-2020** [comment]
Update: https://github.com/Expensify/App/pull/85306#issuecomment-4495750229

**Comment by eh2077** [comment]
@MelvinBot this isn't fixed yet, the issue is still reproducible, please help reopen it

**Comment by MelvinBot** [comment]
🤖 Reopened.

The issue was auto-closed by the `Monthly` staleness sweep on 2026-09-09, not because a fix landed. The original fix ([Expensify/App#82507](https://github.com/Expensify/App/pull/82507)) was reverted by [Expensify/App#85366](https://github.com/Expensify/App/pull/85366) after it caused deploy blockers, so the deploy comment listing this as "deployed to production" is misleading.

The follow-up fix [Expensify/App#85306](https://github.com/Expensify/App/pull/85306) is still open, so the bug remains live. The dependency it was waiting on, [Expensify/App#85825](https://github.com/Expensify/App/pull/85825), has merged.


---
_[view run](https://github.com/Expensify/melvin/actions/runs/34842826377)_ · _no recording available_

**Comment by eh2077** [comment]
hmm, @NJ-2020 seems inactive, @MelvinBot can you try to investigate the issue and try to find a solution? please refer history comments of this issue and PR https://github.com/Expensify/App/pull/85306.

**Comment by MelvinBot** [comment]
## Proposal

### What is the root cause of that problem?

There are **two independent defects**. The thread has been treating them as one, which is why a single FlashList patch keeps looking like the answer and keeps falling short.

**1. You can't start a selection unless you press directly on text.**

[`src/pages/inbox/report/ReportActionItem.tsx:537`](https://github.com/Expensify/App/blob/f4b1457a8668cfe8dc50efef70840ab50c0c031b/src/pages/inbox/report/ReportActionItem.tsx#L537) puts `accessibilityRole={CONST.ROLE.BUTTON}` on the row wrapper for every browser **except Safari**. On web that renders `role="button"` plus `tabindex="0"`, and browsers treat a button-role element as one atomic widget, so a mousedown on the row padding never begins a text selection. Safari was already excluded; Chrome and Chromium browsers were not. This is the part `NJ-2020` confirmed empirically.

**2. Even when selection works, the copied text comes out reversed.**

The chat list is inverted, and on web FlashList implements inversion as a CSS `transform: scaleY(-1)` on the container and on every cell (`PlatformConfig.invertedTransformStyle` in `@shopify/flash-list`'s web `PlatformHelper`). So **visual order is the exact reverse of document order**: index 0 is the newest message and sits at the visual bottom. The browser serialises a selection in document order, so a top-to-bottom drag copies newest → oldest. FlashList's view recycling makes it worse — after any scroll, document order isn't even index order any more.

The merged patch [`011+sort-for-natural-DOM-order`](https://github.com/Expensify/App/blob/f4b1457a8668cfe8dc50efef70840ab50c0c031b/patches/%40shopify/flash-list/%40shopify%2Bflash-list%2B2.3.0%2B011%2Bsort-for-natural-DOM-order.patch) does contain a correct inverted-aware sort, but it is deliberately **opportunistic** — it runs on `focusin` or roughly a second after renders and scrolls settle, and it defers on pointer input specifically to avoid reordering the DOM between `mousedown` and `click`. That deferral is exactly the case a selection drag hits. This matches `sharabai`'s point that the patch was scoped to accessibility and won't cover this.

### What changes do you think we should make in order to solve the problem?

**Fix 1 — stop the button role from blocking selection on web.**

Change the condition at `ReportActionItem.tsx:537` from Safari-only to all pointer-capable web, and keep the button role when a screen reader is active. `isScreenReaderActive` is already in scope at [line 218](https://github.com/Expensify/App/blob/f4b1457a8668cfe8dc50efef70840ab50c0c031b/src/pages/inbox/report/ReportActionItem.tsx#L218), so this costs nothing extra.

This is `NJ-2020`'s [suggested change](https://github.com/Expensify/App/pull/85306#issuecomment-5423100113) plus the screen-reader guard, which keeps the row in the tab order and keeps its button semantics for assistive tech while freeing it for mouse selection.

**Fix 2 — don't rely on DOM order for copy at all; order the clipboard explicitly.**

Add a web-only `copy` listener scoped to the report actions list. On copy: read `window.getSelection()`, resolve each selected row to its list index, sort by that index, build the text, and write it with `event.clipboardData.setData()` + `preventDefault()`.

Two things make this cheap:

- The index map already exists in the DOM. Patch `011` injects a hidden `<div data-flashlist-index={index} aria-hidden>` into every ViewHolder, and ships a `findFocusedIndexFromMarker` walk-up helper that does precisely this node → index lookup. Nothing new to add to the patch.
- App already uses this exact clipboard pattern at [`src/libs/Clipboard/index.ts:42-47`](https://github.com/Expensify/App/blob/f4b1457a8668cfe8dc50efef70840ab50c0c031b/src/libs/Clipboard/index.ts#L42-L47).

This is deterministic regardless of recycling, scroll position, or `scaleY(-1)`. It runs only on copy, so there is no scroll or render cost, no FlashList behaviour change, and **hover is untouched** — which answers the tradeoff `eh2077` flagged.

### What alternative solutions did you explore?

- **Sort the FlashList DOM synchronously (option 2 in the thread).** `sharabai` confirmed this reintroduces stale-hover bugs and was the reason the sort was made opportunistic. Rejected.
- **Remove-and-reinsert instead of recycling (option 3).** Correct but tanks list performance. Rejected.
- **Force a sort on `mousedown`/`selectionchange`.** Reordering the DOM mid-drag clears the browser's active selection, so this would break the drag it is trying to fix. Rejected.

<details>
<summary>What I verified, and what I could not</summary>

Verified from code at `f4b1457`: the `role="button"` condition and its `isSafari()` guard; `role="button"` producing `tabindex="0"` in `react-native-web`'s `createDOMProps`; the chat list using `InvertedFlashList` with `inverted`; web inversion being `transform: scaleY(-1)`; the contents and trigger conditions of patch `011`, including the `data-flashlist-index` marker and the pointer-modality deferral; and the existing `copy`-event clipboard pattern in `Clipboard/index.ts`.

Not verified end-to-end: I could not drive a real drag-selection, clipboard read, or paste in the test harness, so the claim that copy order is still wrong at rest rests on code reading plus the videos already in this thread, not on a fresh reproduction. Worth one manual pass before implementing Fix 2.

</details>

---
**Next Steps for Contributor+ team:**
**To accept:** `@MelvinBot implement [this](https://github.com/Expensify/App/issues/37447#issuecomment-5664017713)` to create a draft PR.
**To refine:** `@MelvinBot <your feedback>`
**To reject:** Explain why you are rejecting Melvin's proposal.

**Comment by neerajbachani** [comment]
Hey @eh2077, I dug into this on desktop web and think I have a clear read on what’s going on. I saw this issue has been open for a long time. Happy to help land a fix if useful.

**What I found:** Multi-message copy on Chrome looks like a clipboard bug, but the browser selection is actually fine. The problem is our own pipeline. Chat rows get `role="button"` on Chrome (Safari skips that on line 537), which turns on `user-select: none` and the selection-scraper hidden marker on each row in `BaseGenericPressable`. When you select across messages, `SelectionScraper` strips those rows from the clone, `getClipboardText` ends up empty, and `setHtml` bails on `if (!text)`. So nothing gets written and paste shows stale clipboard content. Single-message copy often still works because the hidden wrapper isn’t always in the cloned fragment.

**Why Safari is fine:** Same list, same FlashList patch. Safari just never gets the button role, so no suppression and no strip.

**Fix I have ready:** Small opt-in on the chat row: `shouldAllowTextSelection={hasHoverSupport()}`. Keeps `role="button"` for Appium/tab stops but stops the suppression on desktop web. Plus a tiny SelectionScraper tweak to normalize `<button>` → `<div>` for composer paste.

**Branch:** https://github.com/Expensify/App/compare/main...neerajbachani:App:fix/37447-multi-message-copy

**Testing:** Ran this multiple times on both Chrome and Safari: single-message copy, multi-message copy, tab focus on rows. Chrome multi-message was broken before and works after. Safari was already good and still is.

| Chrome | Safari |
|:------:|:------:|
| <video src="https://github.com/user-attachments/assets/93a868aa-a72d-455e-81f8-687c1b201ddd" controls width="100%"></video> | <video src="https://github.com/user-attachments/assets/d4e0b615-fe65-4840-a1a1-4de91b77640b" controls width="100%"></video> |

**Re Melvin’s proposal:** Directionally right that #78527 / button role is involved, but the mechanism isn’t really “atomic button widgets.” It’s our `userSelectNone` + scraper marker chain, and that’s also why the clipboard ends up empty. That part wasn’t traced. Fix 1 (drop role on all web) would hurt tab order unless we add `tabIndex={0}`. Fix 2 (new copy listener + index sort) duplicates existing copy logic and solves ordering we didn’t see broken. Patch 011 already keeps DOM order aligned at rest.

**One follow-up (separate from this fix):** Pasting into the Expensify composer works correctly because `useHtmlPaste` reads the `text/html` flavor and `Parser.htmlToMarkdown` preserves line breaks. But the `text/plain` flavor can still mash messages together with no separators (e.g. `122333444455555`) if you paste into a plain-text target outside the app. That’s existing Safari parity behavior, not introduced by this fix. Worth a separate issue if we want plain-text paste to include line breaks too.

If this matches what others are seeing, I can open a PR or pair with whoever’s already on it. Let me know.

cc: @Julesssss @mallenexpensify

**Comment by Julesssss** [comment]
Hi @neerajbachani thank you for sharing. I'm interested to use your replacement. This is where I think we are at: 

- We fixed the list item ordering issue [here](https://github.com/Expensify/App/pull/82507) but ended up reverting
- The follow-up has been delayed
- We since moved to [FlashList](https://github.com/Expensify/App/pull/85114) which has resolved the issue our follow up fix was working on
- [This](https://github.com/Expensify/App/pull/78527) PR then added buttons to chat rows breaking things again, which we fixed for [Safari](https://github.com/Expensify/App/pull/90541) only

Then your proposal would fix Chrome. It sounds good to me. @eh2077, do you agree with switching to [this](https://github.com/Expensify/App/issues/37447#issuecomment-5668244797) simpler solution?

**Comment by neerajbachani** [comment]
Thanks @Julesssss, that timeline matches what I found too. Happy to open a PR from whenever you and @eh2077 are good with it.

**Comment by eh2077** [comment]
Thanks for your comments! Yes, I think we can go with @neerajbachani 's [solution](https://github.com/Expensify/App/issues/37447#issuecomment-5668244797).
